woocommerce/woocommerce
# Fix: Enhanced Product Collection's Price Range filter by improving currency parsing logic to discard unwanted characters

### Changes proposed in this Pull Request:





This PR addresses an issue in the PriceTextField component, highlighted by @dinhtungdu in [this github comment](https://github.com/woocommerce/woocommerce/issues/45315#issuecomment-1985138515), where the input field would clear itself if users omitted a space between the numeric value and the currency symbol when the currency is configured to appear on the right with a space (e.g., "20$" instead of "20 $"). The problem was rooted in the strict parsing logic that did not account for variations in user input related to currency formatting.

The fix involves an enhancement to the numeric parsing function, which now effectively handles various user inputs regardless of space or placement of the currency symbol. The revised parsing logic uses a regular expression to strip out any characters that are not numeric or the designated decimal separator, and then replaces the locale-specific decimal separator with a period for standard numeric conversion. This approach not only fixes the specific issue but also improves the robustness of the component against similar input variations.

Closes #47351 



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new post
2. Add Product Collection block & choose "Product Catalog" collection
3. Add "Price Range" filter in Inspector controls
4. Test Min & Max inputs of Price Filter. Verify that there is no issue. Here are some example scenarios to test

**Case 1:**
1. In Currency Options, set currency position to "Right with space" as shown in screenshot below:
![image](https://github.com/woocommerce/woocommerce/assets/16707866/95b1a24a-dedc-4cc1-ba42-3717642ca2df)
2. Add any value to "Min" field of Price Range filter, for example 20 and click away. You will notice that it will have a space between value & currency symbol i.e. "20 $". This is as expected. 
3. Now remove the space between value & currency symbol i.e. "20$" and click away. Verify that it should become "20 $".
	- Previously, the input field was getting cleared in this case, which was an issue. 
4. Publish the post and verify the only products within selected Price Range are shown on the Frontend. 

**Case 2:**
Try entering a MIN or MAX price with a decimal, for example, "10.123764". After clicking away, number of decimals should become equal to as set in "Currency Options" as you can see in screenshot above. In my case, It should become "10.12 $"

**Case 3:**
Try entering random values. Here are some inputs and expected outputs:
- 10ABCD -> 10 $
- 10000 -> 10,000 $ (Note, how it's using a Thousand separator from Currency Options setting)
- 1234 KJASHDH -> 1,234 $

As you can see above, it will try to parse numeric values from the input and discard unwanted characters like alphabets, space, etc. 



### Changelog entry




<details>

#### Significance




#### Type




#### Message 
Fix: Enhanced Product Collection's Price Range filter by improving currency parsing logic to discard unwanted characters

#### Comment 

</details>
